### DISCUSSION TITLE
version 3.0.x and SdkClient.close usage

### DISCUSSION THREAD

**Discussion by apetrelli** [awspring/spring-cloud-aws#690] [Q&A] [answered]
Hello everyone.
In AWS SDK v2 I noticed that probably all clients implement interfaces that extend SdkClient, that is AutoCloseable.

I noticed that Spring Cloud Stream 3.0.x does not close these clients, at least not when using it. Probably there is a cleanup when stopping (cannot find it though).

So I am wondering: is it done this way because, after all, it is not necessary to close those clients? In particular I am interested in S3Client.

Thanks in advance
Antonio

**Comment by maciejwalkowiak** [comment, marked as answer]
Hello @apetrelli! If an object that is a Spring bean has a method named "close", Spring automatically calls it when context gets closed, so I think we are good considering that all the clients are Spring beans.

Did you notice that the clients are not getting closed?

**Reply by apetrelli** [reply]
No in fact it is for extra-precaution. In my mind, something that is AutoCloseable means that I have to wrap it into a try-with-resources.

So, again in my mind, it seems like something that I have to create with a builder, put it in a try-with-resources and my code goes into this block.

In fact my question is about the safety of reusing it over and over without the risk of being closed inadvertently by other means (connection drops, for example).

In other words, is it safe to use an instance of S3Client forever?

Thanks again

**Reply by maciejwalkowiak** [reply]
AWS team advices reusing clients infinitely as far as i know and since we don't do anything special, i would say yes it is safe.

**Reply by apetrelli** [reply]
@maciejwalkowiak I simply did not notice it in v2, I noticed that in v1, but then I found it:
https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/best-practices.html#bestpractice1

Thanks a lot!
